### ISSUE TITLE
[question] How to specify 'includedirs'?

### ISSUE THREAD

**Issue by gouriano** [conan-io/conan#8904]
This refers to Windows and Visual Studio. 
I try to change default includedirs of a components on Windows, in package_info() method

I say
self.cpp_info.components["BASE"].includedirs = ['include/Release', 'include']
and get in CONAN_INCLUDE_DIRS: ['include', 'include/Release'] - reverse order!
in conanbuildinfo.cmake, I mean

I say
self.cpp_info.components["BASE"].includedirs = ['include', 'include/Release']
and get in CONAN_INCLUDE_DIRS: ['include', 'include/Release'] - correct, but not what I want

What I really want is
self.cpp_info.components["BASE"].includedirs = ['include/$(Configuration)', 'include']
but in this case I get CONAN_INCLUDE_DIRS: ['include'] - which is totally wrong

**Comment by danimtb** [comment]
Hi @gouriano,

What generator are you using to check the result of CONAN_INCLUDE_DIRS?

**Comment by gouriano** [comment]
It is CMake
I am looking at the contents of conanbuildinfo.cmake,

**Comment by danimtb** [comment]
So not sure what might be happening here but I guess is something related to repeated directories across the dependency graph. CONAN_INCLUDE_DIRS variable aggregates the include directories of all the dependencies in the graph you are installing. As "include" is the value by default, it might be first in the list because it is coming from an upstream dependency.

Another option could be that the generator is picking the value from ``cpp_info.includedirs`` and ``cpp_info.components["BASE"].includedirs``, but I think doing this is not allowed:

```
self.cpp_info.includedirs = []
self.cpp_info.components["BASE"].includedirs = ['include/Release', 'include']
```

Could you give it a go and let me know? thank you

**Comment by gouriano** [comment]
What you suggest cannot work, because:
ConanException: CxxTest/0.1.6 package_info(): self.cpp_info.components cannot be used with self.cpp_info global values at the same time
(Yes, I have verified it)

I have tried few more variants:
When I specify, for example (note that all these directories exist in my package root)
self.cpp_info.components["BASE"].includedirs = ['cmake', 'include/common', 'include']
I get in CONAN_INCLUDE_DIRS = ['include', 'cmake', 'include/common']
'include' always goes first.

When I specify a directory which does not exits in package root:
self.cpp_info.components["BASE"].includedirs = ['notfound', 'include']
it is simply dropped. 
That is not good, I believe. Nonexistent directory is my problem (I write recipe), not yours.
I mean I still want to be able to specify
self.cpp_info.components["BASE"].includedirs = ['include/$(Configuration)', 'include']

Thanks

**Comment by gouriano** [comment]
I have figured out how you can reproduce it.
Create new Conan package:
`conan new hello/0.1`
Change package_info into the following

    def package_info(self):
        self.cpp_info.libs = ["hello"]
        self.cpp_info.includedirs = ['lib', 'include']

run "conan create ."
Then, in a separate directory, create conanfile.txt:

    [requires]
    hello/0.1
    [generators]
    cmake

run "conan install ."
In the generated conanbuildinfo.cmake, CONAN_INCLUDE_DIRS in the correct order: ['lib', 'include']

Now change package_info

    def package_info(self):
        self.cpp_info.components["BASE"].defines.append("_UNICODE")
        self.cpp_info.components["BASE"].includedirs = ['lib', 'include']
        self.cpp_info.components["hellolib"].libs = ["hello"]
        self.cpp_info.components["hellolib"].requires = ["BASE"]

again, "conan create .", then "conan install .", then look into conanbuildinfo.cmake.
CONAN_INCLUDE_DIRS are now in wrong order: [include', 'lib']

So, is it a bug? or I do something wrong?
I use Conan v1.35.2

**Comment by danimtb** [comment]
thanks for the quick response and the clear steps to reproduce the error.
I tried it and I think this is a bug in the way the directories are aggregated when components are used. I have created a tentative PR with the fix at #8927 to see if it does not brake other use cases that might be relying on this behavior.

Thanks a lot for reporting again!

**Comment by gouriano** [comment]
Maybe, once we are at it, it is possible to allow "include/$(Configuration)" ?
that would be useful

**Comment by danimtb** [comment]
@gouriano good news! The fix for this issue has been assigned for the next release. Thanks again for reporting it.

Regarding the "include/$(Configuration)", it is out of the scope of issue but I believe you can conditionally add the directory depending on the build_type already. If that is not the case, please open a new issue to talk about that in particular. Thank you!

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/8927 merged, this will be in next release 1.37
